types/ember/index.d.ts

```diff
@@ -440,27 +440,29 @@ namespace Ember {
     This module implements Observer-friendly Array-like behavior. This mixin is picked up by the
     Array class as well as other controllers, etc. that want to appear to be arrays.
     **/
-    interface Array extends Enumerable {
-        addArrayObserver(target: any, opts?: EnumerableConfigurationOptions): any[];
-        arrayContentDidChange(startIdx: number, removeAmt: number, addAmt: number): any[];
-        arrayContentWillChange(startIdx: number, removeAmt: number, addAmt: number): any[];
-        indexOf(object: any, startAt: number): number;
-        lastIndexOf(object: any, startAt: number): number;
-        objectAt(idx: number): any;
-        objectsAt(...args: number[]): any[];
-        removeArrayObserver(target: any, opts: EnumerableConfigurationOptions): any[];
-        slice(beginIndex?: number, endIndex?: number): any[];
+    interface Array<T> extends Enumerable<T> {
+        addArrayObserver(target: any, opts?: EnumerableConfigurationOptions): this;
+        arrayContentDidChange(startIdx: number, removeAmt: number, addAmt: number): this;
+        arrayContentWillChange(startIdx: number, removeAmt: number, addAmt: number): this;
+        indexOf(object: T, startAt?: number): number;
+        lastIndexOf(object: T, startAt?: number): number;
+        objectAt(idx: number): T;
+        objectsAt(...args: number[]): T[];
+        removeArrayObserver(target: any, opts: EnumerableConfigurationOptions): this;
+        slice(beginIndex?: number, endIndex?: number): T[];
         '@each': EachProxy;
         length: number;
     }
-    const Array: Mixin<Ember.Array>;
+    // Ember.Array rather than Array because the `array-type` lint rule doesn't realize the global is shadowed
+    const Array: Mixin<Ember.Array<any>>;
+
     /**
     An ArrayProxy wraps any other object that implements Ember.Array and/or Ember.MutableArray,
     forwarding all requests. This makes it very useful for a number of binding use cases or other cases
     where being able to swap out the underlying array is useful.
     **/
-    class ArrayProxy extends Object.extend(MutableArray) {
-        content: NativeArray;
+    class ArrayProxy<T> extends Object.extend(MutableArray as Mixin<MutableArray<T>>) {
```

Comment:
Can we do the class/interface merging trick?
```ts
interface ArrayProxy<T> extends MutableArray<T> {}
class ArrayProxy<T> extends Object.extend(MutableArray) {}
```

Reply:
We can, but we'd need to do something like this to avoid a conflicting merge between `MutableArray<T>` and `MutableArray<any>`:

```ts
interface ArrayProxy<T> extends MutableArray<T> {}
class ArrayProxy<T> extends Object.extend(MutableArray as {}) {}
```

Since this is just a declaration file, we could actually get away without the `.extend()` call at all, but real code would need it to pick up the actual mixin implementation.

This is super ugly, but given that type variables apparently aren't in scope in base class expressions, we're pretty restricted. I'm not sure I have a better alternative to propose.

Reply:
Doing a [quick code search on EmberObserver](https://emberobserver.com/code-search?codeQuery=extend%5C\(.*\(Array%7CEnumerable\)&regex=true), code using the enumerable family of mixins only appears in a handful of places, and it tends to be in fairly low-level "plumbing" classes.

Given that, it may not be the end of the world that getting the proper typing takes some gymnastics, but it sure seems unfortunate :/

Reply:
As long as generic mixins aren't a common thing, this should be OK